lib/AST/ASTScopeLookup.cpp

```diff
@@ -503,8 +503,40 @@ GenericTypeOrExtensionWhereOrBodyPortion::computeSelfDC(
   while (i != 0) {
     Optional<NullablePtr<DeclContext>> maybeSelfDC =
         history[--i]->computeSelfDCForParent();
-    if (maybeSelfDC)
-      return *maybeSelfDC;
+    if (maybeSelfDC) {
+      // If we've found a selfDC, we'll definitely be returning something.
+      // However, we may have captured 'self' somewhere down the tree, so we
+      // can't return outright without checking the nested scopes.
+      Optional<NullablePtr<DeclContext>> innerCapturedSelfDC = None;
+      // Start with the innermost scope.
+      size_t j = 0;
+      while (j != i) {
+        // Check if there's a DeclContext for self in this scope. If we're
+        // inside a scope for a completely different type, then we will
+        // encounter a GenericTypeOrExtensionScope somewhere up the tree that
+        // will cause us to forget any DeclContexts we find within.
+        Optional<NullablePtr<DeclContext>> maybeCapturedSelfDC =
+            history[j]->computeSelfDCForParent();
+
+        if (maybeCapturedSelfDC) {
+          // If we haven't already found a captured self context, remember the
+          // first one we see.
+          if (!innerCapturedSelfDC) {
+            innerCapturedSelfDC = maybeCapturedSelfDC;
+          }
+
+          // If there's an intervening scope that causes us to forget the self
+          // context, we should obey that.
+          if ((*maybeCapturedSelfDC).isNull()) {
+            innerCapturedSelfDC = None;
+          }
+        }
+        // Move up to the next scope.
+        j++;
+      }
+
+      return innerCapturedSelfDC ? *innerCapturedSelfDC : *maybeSelfDC;
+    }
```

Comment:
Breaking out into a separate function definitely makes sense, will do!

I'll have to think a bit more on the complexity front--that's a good consideration.

Reply:
Have considered this, and I think if we switch to a top-down approach here rather than a bottom up, we should be able to avoid any quadratic behavior. Currently, if we have a structure like:

```
struct A {
    var x: Int = 0
    func a() {
        struct B {
            func b() {
            // ...
                struct Z {
                    func z() { print(x) }
                }
            // ...
            }
        }
    }
}
```

Then the lookup will, each time we reach a new type scope, traverse the entire history so far (beginning from the local scope of `x`, leading to a quadratic blowup. But if we instead begin looking for captured `self` contexts *at* the type scope and then drill down until we hit something that causes us to forget the `self` context, (i.e., a `GenericTypeOrExtensionScope`), then we'll only re-traverse the scope history at most one additional time total.

You're much more familiar with this code than I am, but let me know if that sounds right to you. I'll bang out an implementation going in that direction.

Reply:
I looked carefully and I agree with you. It's fine. Sorry to have held you up.

Reply:
Not at all--the first implementation you saw *was* quadratic, so it was an important callout. I'll still drop a comment in here to make it clear that the performance has been carefully considered.

Reply:
Great! :)